### PR TITLE
TELCODOCS#1996: Configuring IPsec encryption for multi-node OpenShift clusters

### PR DESCRIPTION
**Pull request by sr1kar99**
Version(s):
4.17

Issue:
[TELCODOCS-1996](https://issues.redhat.com/browse/TELCODOCS-1996)

Link to docs preview:
- [Configuring IPsec encryption for multi-node OpenShift clusters using GitOps ZTP and SiteConfig resources](https://80997--ocpdocs-pr.netlify.app/openshift-enterprise/latest/edge_computing/ztp-deploying-far-edge-sites.html#ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno_ztp-deploying-far-edge-sites)

QE review:
- [x] QE has approved this change.

### CODE REVIEW

**Comment by sr1kar99** [comment]
@sabinaaledort 
Could you please review this PR, and also share the missing steps/TBD info?
Thanks!

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+* You have created a Git repository where you manage your custom site configuration data.
+The repository must be accessible from the hub cluster and be defined as a source repository for the Argo CD application.
+
+* You have installed the `butane` utility, version 0.20.0 or higher.
+
+* You have a PKCS#12 certificate for the IPsec endpoint and a CA cert in PEM format.
+
+* You have installed the NMState Operator
+
+.Procedure
```
Comment (sabinaaledort) [inline modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:29]: Should the first step be extracting the the latest version of the ztp-site-generate container source? Same as in the SNO configuration steps?

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+. Run the `import-certs.sh` script that creates a `MachineConfig` CR to import the external certs:
++
+If the PKCS#12 certificate is protected with a password, set the `-W` argument.
++
+.Example output
+[source,terminal]
+----
+TBD
```
Comment (sabinaaledort) [inline modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:38]: I think this example should be the optional-extra-manifest files example, similar to the one we have for SNO with the additional files in https://github.com/openshift-kni/cnf-features-deploy/tree/master/ztp/gitops-subscriptions/argocd/example/optional-extra-manifest/ipsec
https://github.com/openshift/openshift-docs/pull/80997/files#diff-acb16e551ea95ee44eaeccad594ecbab2254f1c9e6ed9576c975397f77b96c11R75

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+* You have a PKCS#12 certificate for the IPsec endpoint and a CA cert in PEM format.
+
+* You have installed the NMState Operator
+
+.Procedure
+
+. Run the `import-certs.sh` script that creates a `MachineConfig` CR to import the external certs:
++
+If the PKCS#12 certificate is protected with a password, set the `-W` argument.
```
Comment (sabinaaledort) [inline modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:33]: Can you add this line to the SNO steps as well?

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+          name: {{ .metadata.name }}-ipsec-policy
+        spec:
+          nodeSelector:
+            kubernetes.io/hostname: {{ .metadata.name }}
+          desiredState:
+            interfaces:
+            - name: hosta_conn
+              type: ipsec
+              libreswan:
+                left: <cluster_node> <1>
```
Comment (sabinaaledort) [inline modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:78]: left: '%defaultroute', the user can use this setup, can also be updated in the SNO part, https://github.com/openshift-kni/cnf-features-deploy/blob/master/ztp/gitops-subscriptions/argocd/example/optional-extra-manifest/ipsec/ipsec-endpoint-config.yml

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+                ikev2: insist <5>
+                type: tunnel
+----
+<1> Replace `<cluster_node>` with the IP address or DNS hostname of the cluster node for the cluster-side IPsec tunnel.
+<2> Replace `<left_cert>` with the IPsec certificate nickname.
+<3> Replace `<external_host>` with the external host IP address or DNS hostname.
+<4> Replace `<external_address>` with the IP address or subnet of the external host on the other side of the IPsec tunnel.
+<5> Use the IKEv2 VPN encryption protocol only. Do not use IKEv1, which is deprecated.
+
+. Add your `ca.pem` and `left_server.p12` certificates to the `optional-extra-manifest/ipsec` folder.
```
Comment (sabinaaledort) [inline modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:96]: We are pointing here to left_server certificate while in the previous section we suggest replacing it, should we simply point to left_server in the previous section as well? also to follow the official IPsec documentation:  https://docs.openshift.com/container-platform/4.16/networking/network_security/configuring-ipsec-ovn.html#nw-ovn-ipsec-north-south-enable_configuring-ipsec-ovn

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+.. `left_server.p12`: The certificate bundle for the IPsec endpoints
+
+.. `ca.pem`: The certificate authority that you signed your certificates with
+
+. Include the config policy file `ipsec-config-policy.yaml` in the `source-crs` directory in GitOps and reference the file in one of the `PolicyGenerator` CRs.
+
+.Verification
+
+TBD
```
Comment (sabinaaledort) [inline modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:107]: Can be the same as in SNO

**Comment by sabinaaledort** [comment]
Might be missing the custom-manifest folder update part as described in the SNO IPsec config

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc (hunk 3 above)
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:33]: Updated as per the suggestion.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc (hunk 4 above)
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:78]: Updated as per the suggestion.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc (hunk 6 above)
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:107]: Updated as per the suggestion.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc (hunk 1 above)
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:29]: Added this step.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc (hunk 2 above)
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc]: Updated as per the suggestion.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc (hunk 5 above)
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:85]: Updated as per the suggestion.

**Comment by sr1kar99** [comment]
@sabinaaledort 
Updated the content as per the suggestion.
Could you PTAL?
Thanks!

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+* You have configured {rh-rhacm} and the hub cluster for generating the required installation and policy custom resources (CRs) for managed clusters.
+
+* You have created a Git repository where you manage your custom site configuration data.
+The repository must be accessible from the hub cluster and be defined as a source repository for the Argo CD application.
+
+* You have installed the `butane` utility, version 0.20.0 or higher.
+
+* You have a PKCS#12 certificate for the IPsec endpoint and a CA cert in PEM format.
+
+* You have installed the NMState Operator
```
Comment (sabinaaledort) [inline modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:27]: Missing dot in the end of the line

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+.Procedure
+
+. Extract the latest version of the `ztp-site-generate` container source and merge it with your repository where you manage your custom site configuration data.
+
+. Create a file named `ipsec-config-policy.yaml` that contains a node network configuration policy for the NMState Operator:
```
Comment (sabinaaledort) [inline modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:33]: The file will just be there from the previous step, so the user has to just configure the values, not create the file, can be the same as the second step in the SNO procedure

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc (hunk 7 above)
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:27]: Updated as per the suggestion.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc (hunk 8 above)
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc]: Updated as per the suggestion.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 9]
```diff
@@ -0,0 +1,159 @@
+// Module included in the following assemblies:
+//
+// * scalability_and_performance/ztp_far_edge/ztp-advanced-install-ztp.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno_{context}"]
+= Configuring IPsec encryption for multi-node OpenShift clusters using {ztp} and SiteConfig resources
+
+You can enable IPsec encryption in managed multi-node clusters that you install using {ztp} and {rh-rhacm-first}.
+You can encrypt external traffic between pods and IPsec endpoints external to the managed cluster. All pod-to-pod network traffic between nodes on the OVN-Kubernetes cluster network is encrypted with IPsec in Transport mode.
```
Comment (bradyallenjohnson) [inline modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:10]: These 2 sentences incorrectly identify the type of traffic to be encrypted. I would reword them as follows:
(You may even consider mentioning North/South traffic which is OCP cluster ingress/egress traffic as opposed to East/West traffic which is OCP cluster intra node traffic)
You can encrypt traffic between the managed cluster and IPsec endpoints external to the managed cluster.
All network traffic between nodes on the OVN-Kubernetes cluster network is encrypted with IPsec in Transport mode.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+            kubernetes.io/hostname: {{ .metadata.name }}
+          desiredState:
+            interfaces:
+            - name: hosta_conn
+              type: ipsec
+              libreswan:
+                left: '%defaultroute'
+                leftid: '%fromcert'
+                leftmodecfgclient: false
+                leftcert: left_server
```
Comment (bradyallenjohnson) [inline modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:72]: You should mention that this name "left_server" should match the name of the certificate used on the remote system.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+                leftrsasigkey: '%cert'
+                right: <external_host> <1>
+                rightid: '%fromcert'
+                rightrsasigkey: '%cert'
+                rightsubnet: <external_address> <2>
+                ikev2: insist <3>
+                type: tunnel
+----
+<1> Replace `<external_host>` with the external host IP address or DNS hostname.
+<2> Replace `<external_address>` with the IP address or subnet of the external host on the other side of the IPsec tunnel.
```
Comment (bradyallenjohnson) [inline modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:82]: Im pretty sure this can only be a subnet, not an individual IP address. If it is only 1 address, then it should be something like 192.168.100.20/32.
As such, this should be reworded as follows:
"with the IP subnet of the external host on the other side of the IPsec tunnel."

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig.adoc [hunk 12]
```diff
@@ -44,23 +41,21 @@ interfaces:
...
     rightid: '%fromcert'
     rightrsasigkey: '%cert'
-    rightsubnet: <external_address> <4>
-    ikev2: insist <5>
+    rightsubnet: <external_address> <2>
+    ikev2: insist <3>
     type: tunnel
 ----
-<1> Replace `<cluster_node>` with the IP address or DNS hostname of the cluster node for the cluster-side IPsec tunnel.
-<2> Replace `<left_cert>` with the IPsec certificate nickname.
```
Comment (bradyallenjohnson) [inline modules/ztp-configuring-ipsec-using-ztp-and-siteconfig.adoc:60]: This is still valid and should remain.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig.adoc [hunk 13]
```diff
@@ -44,23 +41,21 @@ interfaces:
...
-<3> Replace `<external_host>` with the external host IP address or DNS hostname.
-<4> Replace `<external_address>` with the IP address or subnet of the external host on the other side of the IPsec tunnel.
-<5> Use the IKEv2 VPN encryption protocol only. Do not use IKEv1, which is deprecated.
+<1> Replace `<external_host>` with the external host IP address or DNS hostname.
+<2> Replace `<external_address>` with the IP address or subnet of the external host on the other side of the IPsec tunnel.
```
Comment (bradyallenjohnson) [inline modules/ztp-configuring-ipsec-using-ztp-and-siteconfig.adoc:57]: Same comment as before about the subnet.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc (hunk 11 above)
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:82]: Updated as per the suggestion.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig.adoc (hunk 13 above)
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig.adoc:57]: Updated as per the suggestion.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc (hunk 10 above)
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:72]: Updated as per the suggestion.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig.adoc (hunk 12 above)
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig.adoc:60]: These updates were made based on the following suggestions from @sabinaaledort 

- https://github.com/openshift/openshift-docs/pull/80997#discussion_r1764913489
- https://github.com/openshift/openshift-docs/pull/80997#discussion_r1764907052


Could you please confirm if these updates are fine? @bradyallenjohnson

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc (hunk 9 above)
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc]: Updated as per the suggestion.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig.adoc (hunk 12 above)
Comment (bradyallenjohnson) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig.adoc:60]: Ok, as long as you suggest the left_cert field should match, its fine.

File: modules/ztp-verifying-ipsec.adoc [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+.Procedure
+
+. Start a debug pod for the managed cluster by running the following command:
++
+[source,terminal]
+----
+$ oc debug node/<node_name>
+----
+
+. Check that the IPsec policy is applied in the cluster node:
```
Comment (skopacz1) [inline modules/ztp-verifying-ipsec.adoc:28]: I think that even if these commands are being run from a shell as root, the [docs guidelines rule](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#procedure-step-that-introduces-a-command) about explicitly mentioning commands still applies:
```suggestion
. Check that the IPsec policy is applied in the cluster node by running the following command:
```

File: modules/ztp-verifying-ipsec.adoc [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+  dir fwd priority 1757377 ptype main
+  tmpl src 10.1.232.10 dst 10.1.28.190
+    proto esp reqid 16393 mode tunnel
+src 10.1.232.10/32 dst 172.16.123.0/24
+  dir in priority 1757377 ptype main
+  tmpl src 10.1.232.10 dst 10.1.28.190
+    proto esp reqid 16393 mode tunnel
+----
+
+. Check that the IPsec tunnel is up and connected:
```
Comment (skopacz1) [inline modules/ztp-verifying-ipsec.adoc:52]: Same comment as before:
```suggestion
. Check that the IPsec tunnel is up and connected by running the following command:
```

File: modules/ztp-verifying-ipsec.adoc [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+  replay-window 0 flag af-unspec esn
+  auth-trunc hmac(sha1) 0xd960ddc0a6baaccb343396a51295e08cfd8aaddd 96
+  enc cbc(aes) 0x0273c02e05b4216d5e652de3fc9b3528fea94648bc2b88fa01139fdf0beb27ab
+  anti-replay esn context:
+   seq-hi 0x0, seq 0x0, oseq-hi 0x0, oseq 0x0
+   replay_window 128, bitmap-length 4
+   00000000 00000000 00000000 00000000
+----
+
+. Ping a known IP in the external host subnet.
```
Comment (skopacz1) [inline modules/ztp-verifying-ipsec.adoc:82]: Same comment as before:
```suggestion
. Ping a known IP in the external host subnet by running the following command.
```

File: modules/ztp-verifying-ipsec.adoc [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+For example, ping an IP in the `rightsubnet` range that you set in `ipsec/ipsec-endpoint-config.yaml`:
```
Comment (skopacz1) [inline modules/ztp-verifying-ipsec.adoc:83]: I think this sentence should add some descriptive nouns for clarity and formality:
```suggestion
For example, ping an IP address in the `rightsubnet` range that you set in the `ipsec/ipsec-endpoint-config.yaml` file:
```

File: modules/ztp-verifying-ipsec.adoc [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
++
+[source,terminal]
+----
+sh-5.1# ping 172.16.110.8
+----
++
+.Example output
+[source,terminal]
+----
+sh-5.1# ping 172.16.110.8
```
Comment (skopacz1) [inline modules/ztp-verifying-ipsec.adoc:93]: If this is example output, only the output should be included:
```suggestion
```

**Review by skopacz1 — COMMENTED** [review]
Left some comments, let me know if you have any questions. Otherwise solid work!

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 19]
```diff
@@ -0,0 +1,160 @@
+// Module included in the following assemblies:
+//
+// * scalability_and_performance/ztp_far_edge/ztp-advanced-install-ztp.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno_{context}"]
+= Configuring IPsec encryption for multi-node OpenShift clusters using {ztp} and SiteConfig resources
```
[the remark this answers is not in the document]
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:7]: Updated as per the suggestions.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+* You have installed the OpenShift CLI (`oc`).
+
+* You have logged in to the hub cluster as a user with `cluster-admin` privileges.
+
+* You have configured {rh-rhacm} and the hub cluster for generating the required installation and policy custom resources (CRs) for managed clusters.
+
+* You have created a Git repository where you manage your custom site configuration data.
+The repository must be accessible from the hub cluster and be defined as a source repository for the Argo CD application.
+
+* You have installed the `butane` utility, version 0.20.0 or higher.
```
[the remark this answers is not in the document]
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:23]: Updated as per the suggestion.

[the remark this answers is not in the document]
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:23]: Updated as per the suggestion.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+
+* You have installed the NMState Operator.
+
+.Procedure
+
+. Extract the latest version of the `ztp-site-generate` container source and merge it with your repository where you manage your custom site configuration data.
+
+. Configure the `optional-extra-manifest/ipsec/ipsec-config-policy.yaml` file with the required values that configure IPsec in the cluster. For example:
++
+.Sample `ConfigurationPolicy` object for creating an IPsec configuration
```
[the remark this answers is not in the document]
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:35]: Updated as per the suggestion.

Will add sample values post GA.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+<2> Replace `<external_host>` with the external host IP address or DNS hostname.
+<3> Replace `<external_address>` with the IP subnet of the external host on the other side of the IPsec tunnel.
+<4> Use the IKEv2 VPN encryption protocol only. Do not use IKEv1, which is deprecated.
+
+. Add your `ca.pem` and `left_server.p12` certificates to the `optional-extra-manifest/ipsec` folder.
+The certificate files are required for the Network Security Services (NSS) database on each host. These files are imported as part of the Butane configuration in later steps.
+
+.. `left_server.p12`: The certificate bundle for the IPsec endpoints
+
+.. `ca.pem`: The certificate authority that you signed your certificates with
```
[the remark this answers is not in the document]
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:91]: Updated it as follows:

. Add the following certificates to the `optional-extra-manifest/ipsec` folder:

** `left_server.p12`: The certificate bundle for the IPsec endpoints

** `ca.pem`: The certificate authority that you signed your certificates with
+
The certificate files are required for the Network Security Services (NSS) database on each host. These files are imported as part of the Butane configuration in later steps.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+
+. Open a shell prompt at the `optional-extra-manifest/ipsec` folder of the Git repository where you maintain your custom site configuration data.
+
+. Run the `optional-extra-manifest/ipsec/import-certs.sh` script to generate the required Butane and `MachineConfig` CRs to import the external certs:
```
[the remark this answers is not in the document]
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:95]: Updated as per the suggestion.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+      └── example
+           └── optional-extra-manifest
+                └── ipsec
+                     ├── 99-ipsec-master-import-certs.bu <1> 
+                     ├── 99-ipsec-master-import-certs.yaml
+                     ├── 99-ipsec-worker-import-certs.bu
+                     ├── 99-ipsec-worker-import-certs.yaml
+                     ├── import-certs.sh
+                     ├── ca.pem <2>
+                     ├── left_server.p12
```
[the remark this answers is not in the document]
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:113]: Updated as per the suggestion.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+                     ├── enable-ipsec.yaml
+                     ├── ipsec-config-policy.yaml
+                     └── README.md
+----
+<1> The `ipsec/import-certs.sh` script generates the Butane and endpoint configuration CRs.
+<2> Add the `ca.pem` and `left_server.p12` certificate files that are relevant to your network.
+
+. Create a `custom-manifest/` folder in the repository where you manage your custom site configuration data.
+Add the `enable-ipsec.yaml` and `99-ipsec-*` YAML files to the directory.
```
[the remark this answers is not in the document]
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:122]: Updated as per the suggestion.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 26]
```diff
@@ -0,0 +1,160 @@
...
+For example:
++
```
[the remark this answers is not in the document]
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:124]: Updated as per the suggestion.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 27]
```diff
@@ -0,0 +1,160 @@
...
+  ├── site1-mno-du.yaml
+  ├── extra-manifest/
+  └── custom-manifest
+        ├── enable-ipsec.yaml
+        ├── 99-ipsec-master-import-certs.yaml
+        └── 99-ipsec-worker-import-certs.yaml
+----
+
+. In your `SiteConfig` CR, add the `custom-manifest/` directory to the `extraManifests.searchPaths` field.
+For example:
```
[the remark this answers is not in the document]
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:137]: Updated as per the suggestion.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 28]
```diff
@@ -0,0 +1,160 @@
...
+clusters:
+- clusterName: "site1-mno-du"
+  networkType: "OVNKubernetes"
+  extraManifests:
+    searchPaths:
+      - extra-manifest/
+      - custom-manifest/
+----
+
+. Include the config policy file `ipsec-config-policy.yaml` in the `source-crs` directory in GitOps and reference the file in one of the `PolicyGenerator` CRs.
```
[the remark this answers is not in the document]
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:150]: Updated as per the suggestion.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc [hunk 29]
```diff
@@ -0,0 +1,160 @@
...
+
+. Commit the `SiteConfig` CR changes and updated files in your Git repository and push the changes to provision the managed cluster and configure IPsec encryption.
++
+The Argo CD pipeline detects the changes and begins the managed cluster deployment.
++
+During cluster provisioning, the {ztp} pipeline appends the CRs in the `/custom-manifest` directory to the default set of extra manifests stored in `extra-manifest/`.
```
[the remark this answers is not in the document]
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:156]: Updated as per the suggestion.

[the remark this answers is not in the document]
Comment (sr1kar99) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc:156]: You're right! It should come after.

File: modules/ztp-verifying-ipsec.adoc [hunk 30]
```diff
@@ -0,0 +1,97 @@
+// Module included in the following assemblies:
+//
+// * scalability_and_performance/ztp_far_edge/ztp-advanced-install-ztp.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="ztp-verifying-ipsec_{context}"]
+= Verifying the IPsec encryption
+
+You can verify that the IPsec encryption is successfully applied in a managed OpenShift cluster.
```
[the remark this answers is not in the document]
Comment (sr1kar99) [reply modules/ztp-verifying-ipsec.adoc:9]: Updated as per the suggestion.

File: modules/ztp-verifying-ipsec.adoc (hunk 14 above)
Comment (sr1kar99) [reply modules/ztp-verifying-ipsec.adoc:28]: Updated as per the suggestion.

File: modules/ztp-verifying-ipsec.adoc (hunk 15 above)
Comment (sr1kar99) [reply modules/ztp-verifying-ipsec.adoc:52]: Updated as per the suggestion.

File: modules/ztp-verifying-ipsec.adoc (hunk 17 above)
Comment (sr1kar99) [reply modules/ztp-verifying-ipsec.adoc:83]: Updated as per the suggestion.

File: modules/ztp-verifying-ipsec.adoc (hunk 18 above)
Comment (sr1kar99) [reply modules/ztp-verifying-ipsec.adoc:93]: Updated as per the suggestion.

File: modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc (hunk 22 above)
[the remark this answers is not in the document]
Comment (skopacz1) [reply modules/ztp-configuring-ipsec-using-ztp-and-siteconfig-for-mno.adoc]: Nice, I like that solution!

**Review by sheriff-rh — APPROVED** [review]
This is excellent work! Merging to 4.17+.
